server/src/main/java/org/elasticsearch/index/query/MatchQueryBuilder.java

```diff
@@ -81,10 +81,16 @@ public class MatchQueryBuilder extends AbstractQueryBuilder<MatchQueryBuilder> {
 
     private boolean autoGenerateSynonymsPhraseQuery = true;
 
+    /**
+     * Indicates that this MatchQueryBuilder has already been intercepted and rewritten,
+     * so subsequent rewrite rounds can short-circuit interception. 
+     */
+    private final boolean interceptedAndRewritten;
```

Comment:
I don't understand the purpose of this. If it was a no-op, and thus returned the same query, rewriting will cease (because we rewrite until the same thing is returned), or if it was rewritten and modified, we are returning a separate object.

Reply:
The issue is that if it's a non inference field, we'll return another match query as part of the boolean query clause. I tried to not have this, but I couldn't find a better way of avoiding the too many rewrite rounds issue.

Reply:
> The issue is that if it's a non inference field, we'll return another match query as part of the boolean query clause. I tried to not have this, but I couldn't find a better way of avoiding the too many rewrite rounds issue.

This is a bug. This tells me that we are attempting to rewrite and are mutating the query even when there is nothing to do.

Reply:
I don't think it's a bug. This is how we know it's OK to short-circuit the rewrite. 

Let's say we have this match query: 
```
GET my-index-*/_search
{
  "query": {
    "match": {
      "my_field": "test"
    }
  }
}
```

And `my_field` is different types in different indices - one is a `semantic_text` field and one is just plain `text`. 

We rewrite this to a boolean query that's essentially: 
- SHOULD 
  - MUST index = the index with the text field 
  - MUST match(my_field, test)
- SHOULD 
  - MUST index = the index with the semantic_text field 
  - MUST semantic(my_field, test)

Now, when we rewrite the boolean query, we get to that should clause, and if we don't know to short circuit, we will attempt to rewrite it again. This is basically a flag that says it's already all good. 

I tried adding it in the interceptor itself, but that doesn't work as hoped. 
I don't want to create a "NonInterceptedMatchQuery" subclass, but that's the only other thing I can think of if the boolean isn't acceptable. That said, I think the Boolean is OK for this use case?

Reply:
I am saying, the rewrite logic should be able to handle a no-op without setting any flags on any query. Treating semantic as special in this way is a very bad smell and indicates something wrong with how we are doing this.

The rewrite logic is flawed in some way.

> Now, when we rewrite the boolean query, we get to that should clause, and if we don't know to short circuit, we will attempt to rewrite it again.

I understand that this is the bug, as we will try to rewrite `SHOULD.MUST.match(my_field, test)` again because we see that my_field is semantic in the `semantic_index` even though this is handled in the separate boolean clause. 

I don't immediately know the answer here, but this is a bad smell and will lead to bugs.

Reply:
I'll experiment with this a bit more today, possibly with wrapping the query after it has been intercepted.

Reply:
Seems like we need a `IndexRestrictedQueryBuilder` to encapsulate this, which isn't available for users in the API but we can use internally.

Reply:
OK, updated in [b97d148](https://github.com/elastic/elasticsearch/pull/117839/commits/b97d148710f53081aa920c5ebcb21b820d4ca372)

Reply:
Resolving this thread as we removed the boolean in `MatchQueryBuilder`